### PR TITLE
Update for editions

### PR DESCRIPTION
**Pull request by jchadwick-buf**
Updates protovalidate, adds support for new ignore semantics. The PR is huge but it's mostly gencode; the actual CEL code wasn't too substantially altered.

~~Unfortunately I could not get the tests to pass 100%. There is a minor wrinkle where, I _believe_, the behavior of `has(x.y)` and `x.y` needs to differ (specifically with proto3 optional) but it can not due to the way the interpreter works. It _seems_ like this only impacts some rather small edge cases so I guess we might be able to live with it in the interim. I've just added said tests to the nonconforming list.~~ **Update:** The issues still exist upstream (assuming the issue is valid and not a mistake on my part) but we are now working around it, and tests now pass 100% (sans the existing failures.)

All other tests pass and I _believe_ my fixes are correct. The e-mail check was corrected to not support bracketed emails as was done with the Go version.

### CODE REVIEW

File: gen/buf/validate/conformance/cases/bool_pb2.py [hunk 1]
```diff
@@ -14,12 +14,23 @@
 
 # -*- coding: utf-8 -*-
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
+# NO CHECKED-IN PROTOBUF GENCODE
```
Comment (rodaine) [inline gen/buf/validate/conformance/cases/bool_pb2.py:17]: ?

File: tests/conformance/nonconforming.yaml [hunk 2]
```diff
@@ -6,3 +6,11 @@ standard_constraints/well_known_types/duration:
 standard_constraints/well_known_types/timestamp:
  - gte_lte/invalid/above
  - lte/invalid
+# There are a couple of cases where celpy's mapping of has(...) and x.y seem to
+# be incongruent with protobuf presence, resulting in a runtime error instead of
+# a validation error.
```
Comment (rodaine) [inline tests/conformance/nonconforming.yaml:11]: Can we report this upstream?

**Review by rodaine — APPROVED** [review]

File: protovalidate/internal/extra_func.py [hunk 3]
```diff
@@ -49,9 +50,14 @@ def _validate_hostname(host):
 
 
 def validate_email(addr):
-    if "<" in addr and ">" in addr:
-        addr = addr.split("<")[1].split(">")[0]
+    addrs = getaddresses([addr])
```
Comment (stefanvanburen) [inline protovalidate/internal/extra_func.py:53]: would [`parseaddr`](https://docs.python.org/3/library/email.utils.html#email.utils.parseaddr) be more appropriate for a single addr?

File: tests/conformance/nonconforming.yaml (hunk 2 above)
Comment (jchadwick-buf) [reply tests/conformance/nonconforming.yaml:11]: I really need to do more investigation into the problem in order to understand what's going on, but I haven't had time. It's complicated slightly by the fact that we don't even use celpy's protobuf integration afaict, we just use MapTypes manually. I tried to use celpy's MessageType just to see if that would save us some trouble, but it had even more issues, so I think there's good reason for what we did.

File: protovalidate/internal/extra_func.py (hunk 3 above)
Comment (jchadwick-buf) [reply protovalidate/internal/extra_func.py:53]: Actually yeah, it _should_ be safe, due to the equality check. I originally switched to `getaddresses` because `parseaddr` succeeds just fine if you pass a group instead of a single address.

File: gen/buf/validate/conformance/cases/bool_pb2.py (hunk 1 above)
Comment (jchadwick-buf) [reply gen/buf/validate/conformance/cases/bool_pb2.py:17]: No idea.

https://github.com/protocolbuffers/protobuf/pull/16226

Probably used to prevent gencode in Piper, or something.

File: protovalidate/internal/extra_func.py (hunk 3 above)
Comment (jchadwick-buf) [reply protovalidate/internal/extra_func.py]: Done.

File: protovalidate/internal/constraints.py [hunk 4]
```diff
@@ -269,24 +279,66 @@ def check_field_type(field: descriptor.FieldDescriptor, expected: int, wrapper_n
...
+        return message
+    elif field.message_type is not None and field.label != descriptor.FieldDescriptor.LABEL_REPEATED:
+        return message_factory.GetMessageClass(field.message_type)()
+    else:
+        return field.default_value
+
+
+def _proto_equals(a, b):
+    if isinstance(a, message.Message) and isinstance(b, message.Message):
+        # TODO: Python protobuf does not have the ability to compare messages.
```
Comment (jchadwick-buf) [inline protovalidate/internal/constraints.py:307]: I just had a thought. While Python protobuf doesn't have a comparison function, celpy does. Maybe it would be better to convert the zero value to a celpy type: we are already calling `_field_value_to_cel` right after the zero check anyways. I'll try that next week.

Comment (jchadwick-buf) [reply protovalidate/internal/constraints.py]: This works and is now committed, removing `_proto_equals` entirely.

File: tests/conformance/nonconforming.yaml (hunk 2 above)
Comment (jchadwick-buf) [reply tests/conformance/nonconforming.yaml]: It really feels tricky to debug this but after probing the Go implementation and carefully reading the cel-python source code I think I have a good enough understanding of the problem. In order to confirm my understanding, I implemented a workaround in ab874b27d746fd5f1ab3b26f820a2f358ded42c0.

The workaround trivially passes our conformance tests and while the mechanism used to implement it is ugly, the actual logic is fairly simple and straight-forward. I committed the workaround thunks as their own commit so that we can drop this commit if we decide it's too ugly to include in protovalidate-python, but I hope by actually producing the workaround we can have a bit more confidence that we actually do understand the nature of the problem correctly. It's possible there are other issues lurking in the handling of presence and defaults (I think the conformance test suite needs to grow a lot...) but at the very least I think this covers the issue we're currently facing correctly and completely.

I reported the issue upstream, albeit without a minimal reproduction for now. I think it will at least need a minimal reproduction, but I already spent a lot more time on this issue than intended, so I might need to loop back to it after continuing onto other protovalidate implementations.

https://github.com/cloud-custodian/cel-python/issues/73

Comment (rodaine) [reply tests/conformance/nonconforming.yaml]: I don't mind the patch, assuming it will continue to work for versions of cel-python that fix this particular code. Should we pin pv-python to be more tightly tied to these versions of cel-python?

Comment (jchadwick-buf) [reply tests/conformance/nonconforming.yaml]: Yeah, I think we definitely should do that.

I'm hopeful that it can be fixed upstream, but it may be a "PRs welcome" type of situation, and if so I'll need to try picking up some lark-parser knowledge.

**Comment by jchadwick-buf** [comment]
OK, I think all outstanding issues should be resolved. I don't think we're in a rush to close this though so I'll leave some time for people to respond before merging this.
